[chore] improve binary_numeric conformance

lwwmanning opened the pull request:
⏺ Summary of Work Completed

  Based on the git diff, here's a comprehensive summary of the changes made:

  1. Enhanced Binary Numeric Conformance Testing

  - Modified the test harness in /vortex-array/src/compute/conformance/binary_numeric.rs to handle overflow/underflow on a per-element basis rather
  than skipping entire operations
  - Added detailed documentation explaining the test strategy and supported operations
  - Implemented per-element overflow tracking that:
    - Runs operations even when some elements would overflow
    - Validates non-overflowing elements produce correct results
    - Verifies that elements expected to overflow actually do fail the checked operation
  - Added edge case testing for special values (zero, max, min, NaN, infinity for floats)

  2. Added Binary Numeric Tests Across Multiple Encodings

  Added comprehensive binary numeric conformance tests to 11 different array encodings:

  - ALP (Adaptive Lossless Floating-Point): Tests for f32/f64 arrays
  - ALP-RD (ALP with Reference Data): Tests for f32/f64 arrays with reference data encoding
  - Dict (Dictionary): Tests for dictionary-encoded arrays including nullable and sliced variants
  - BitPacking: Tests for bit-packed integer arrays (u8, u16, u32, u64, i32)
  - Delta: Tests for delta-encoded arrays with constant values
  - FoR (Frame of Reference): Tests for frame-of-reference encoded arrays
  - RunEnd: Tests for run-end encoded arrays with repeated values
  - Sparse: Tests for sparse arrays with various fill values
  - ZigZag: Tests for zigzag-encoded signed integers
  - Primitive: Tests for basic primitive arrays
  - Various consistency tests were also added alongside the binary numeric tests

  3. Test Case Improvements

  - Simplified test cases for Delta encoding to use arrays with constant values (e.g., [1, 1, 1, 1, 1]) which better test the encoding's behavior
  - Added diverse test scenarios including:
    - Basic small arrays
    - Large arrays (100+ elements)
    - Single element arrays
    - Nullable arrays
    - Arrays with specific patterns (alternating values, repeated values)

The overall effect is a much more robust and comprehensive test suite that ensures binary numeric operations behave correctly across all array encodings, even in edge cases involving overflow/underflow.

lwwmanning committed:
binary

Signed-off-by: Will Manning <will@willmanning.io>

lwwmanning committed:
binary

Signed-off-by: Will Manning <will@willmanning.io>

lwwmanning committed:
binary

Signed-off-by: Will Manning <will@willmanning.io>

lwwmanning committed:
wip

Signed-off-by: Will Manning <will@willmanning.io>

lwwmanning committed:
wip

Signed-off-by: Will Manning <will@willmanning.io>

lwwmanning committed:
wip

Signed-off-by: Will Manning <will@willmanning.io>

lwwmanning committed:
wip

Signed-off-by: Will Manning <will@willmanning.io>

lwwmanning committed:
wip

Signed-off-by: Will Manning <will@willmanning.io>

lwwmanning committed:
wip

Signed-off-by: Will Manning <will@willmanning.io>

lwwmanning committed:
wip

Signed-off-by: Will Manning <will@willmanning.io>

lwwmanning commented:
@claude can you review this PR please?

lwwmanning committed:
wip

Signed-off-by: Will Manning <will@willmanning.io>

danking commented on vortex-array/src/compute/conformance/binary_numeric.rs:
```diff
@@ -41,44 +102,333 @@ where
+                let original_value = &original_values[idx];
+                let overflow_check = original_value
+                    .as_primitive()
+                    .checked_binary_numeric(&scalar_one.as_primitive(), operator);
```
Unless I'm dense, this is re-running exactly the code from 124-127 on the exact same values.

danking commented on vortex-array/src/compute/conformance/binary_numeric.rs:
```diff
@@ -41,44 +102,333 @@ where
+                    array.encoding_id(),
+                    idx,
+                );
+            } else {
```
same here

danking commented on vortex-array/src/compute/conformance/binary_numeric.rs:
```diff
@@ -41,44 +102,333 @@ where
+    }
+}
+
+fn test_binary_numeric_edge_cases_for_type<T>(array: ArrayRef)
```
```suggestion
fn test_binary_numeric_edge_cases_signed<T>(array: ArrayRef)
```

danking commented on vortex-array/src/compute/conformance/binary_numeric.rs:
```diff
@@ -41,44 +102,333 @@ where
+    test_binary_numeric_with_scalar(array.clone(), T::max_value());
+
+    // Test with min value (0 for unsigned)
+    test_binary_numeric_with_scalar(array, T::min_value());
```
this just duplicates the first check on line 312

danking commented on vortex-array/src/compute/conformance/binary_numeric.rs:
```diff
@@ -41,44 +102,333 @@ where
+    // Test with special float values (NaN, Infinity)
+    test_binary_numeric_with_scalar(array.clone(), T::nan());
+    test_binary_numeric_with_scalar(array.clone(), T::infinity());
+    test_binary_numeric_with_scalar(array, T::neg_infinity());
```
If we're enumerating all the funny guys, it seems we should include `T::min_positive_value()` which is a subnormal floating-point number

danking commented on vortex-array/src/compute/conformance/binary_numeric.rs:
```diff
@@ -19,11 +41,50 @@ fn to_vec_of_scalar(array: &dyn Array) -> Vec<Scalar> {
+
+    // Then test edge cases if we have enough data
+    if array.len() >= 5 {
+        test_binary_numeric_edge_cases(array);
```
I don't think any of the tests in here rely on the array being greater than or equal to 5.

danking commented on vortex-array/src/compute/conformance/binary_numeric.rs:
```diff
@@ -41,44 +102,333 @@ where
+                    scalar_value,
+                );
+            } else {
+                // For overflow elements, verify that the operation would indeed overflow
```
same point here about overflow: I'm pretty sure we're re-running the command on the same value from original_values that caused us to take this branch.

danking reviewed (changes requested):
little things

lwwmanning commented:
@claude can you address the code review comments?

claude[bot] committed:
Fix binary numeric conformance test issues per code review

- Remove duplicate overflow checks that re-ran already computed operations
- Rename test_binary_numeric_edge_cases_for_type to test_binary_numeric_edge_cases_signed for clarity
- Remove duplicate T::min_value() test for unsigned types (same as T::zero())
- Add T::min_positive_value() test for float types to include subnormal numbers
- Remove unnecessary array length constraint (>= 5) for edge case testing

Co-authored-by: Will Manning <lwwmanning@users.noreply.github.com>

danking reviewed (approved):
lgtm
